Improve documentation for gpg subkey creation

glensc opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | yes/no
| BC Break      | yes/no
| New Feature   | yes/no
| RFC           | yes/no
| QA            | yes/no

### Description

This updates GPG section to match my system (`gpg (GnuPG) 2.2.25` from [brew]).

[brew]: https://brew.sh/

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

glensc committed:
Add `gpg --list-keys` tip for finding existing GPG keys to `README.md`

Signed-off-by: Elan Ruusamäe <glen@pld-linux.org>
Signed-off-by: Marco Pivetta <ocramius@gmail.com>

glensc commented:
> Type `addkey` and select signing or s for capabilities. RSA key type is recommended for greatest compatibility. Type save to persist the new subkey to your master key. Make a note of the Key ID as you will need it in the next step.


I'm confused here, "s" is supposed to be key to type? i have only numeric answers:

```
gpg> addkey
Please select what kind of key you want:
   (3) DSA (sign only)
   (4) RSA (sign only)
   (5) Elgamal (encrypt only)
   (6) RSA (encrypt only)
  (14) Existing key from card
Your selection? s
Invalid selection.
```

glensc commented:
Next, I also get asked about bits:

```
RSA keys may be between 1024 and 4096 bits long.
What keysize do you want? (3072)
```

So what's the recommendation here for git signing in github? 

From  https://en.wikipedia.org/wiki/Key_size I read:
> Since 2015, NIST recommends a minimum of 2048-bit keys for RSA, an update to the widely-accepted recommendation of a 1024-bit minimum since at least 2002.

glensc committed:
Write out clearly what is the current gpg key recommendation

Signed-off-by: Elan Ruusamäe <glen@pld-linux.org>

glensc committed:
Add a note about key expiry setup

Signed-off-by: Elan Ruusamäe <glen@pld-linux.org>

Ocramius commented:
```
Your selection? s
Invalid selection.
```

Looks like something that has changed on newer releases: certainly want a key with signing capabilities (used to be `s` for `sign`, `e` for `encrypt` and something else for certification/authentication)

 > So what's the recommendation here for git signing in github?

The default of 3072 works: bigger values lead to some overhead, but overall not noticeable. My old key is still 2048, for example.

glensc commented:
I've added an example commit with 2048, change to 3072?

glensc committed:
Use relative path for gpg ephemeral homedir for security

This may be paranoid, but also avoids unwanted surprises of multi-user systems.

Signed-off-by: Elan Ruusamäe <glen@pld-linux.org>

Ocramius commented:
Probably better to keep it the default suggested one: that's really up to GnuPG to endorse, while we're mostly only asking for a key, rather than endorsing values to configure one (security-sensitive stuff that we shouldn't touch)

glensc commented:
If I attempt to remove password, I end up with a confusing error:

```
gpg> passwd
gpg: key XXXX/XXXX: error changing passphrase: No secret key

gpg> save
Key not changed so no update needed.
```

but in reality password was removed after `passwd` command, and yes, the `save` is unneeded.

glensc committed:
Add note about password remove giving error

Signed-off-by: Elan Ruusamäe <glen@pld-linux.org>

Ocramius commented on README.md:
````diff
@@ -40,13 +40,16 @@ in the GitHub Docs.
 ```
 
-Type `addkey` and select signing or s for capabilities. RSA key type is recommended for greatest compatibility.
+A signing key type RSA with 3072 bits is recommended for greatest compatibility.
````
Discussed during other comments in the review: I'd say that the length should not be endorsed by us, since the tool already endorses one, while if we endorse it here it will become old knowledge pretty quickly.

glensc replied:
I thought to express it's as a recommendation as is RSA recommended here. Would "at least 3072" be better?

There's (in this PR changes) also a recommendation about expiry, which addresses that the security related recommendation can become "old knowledge" pretty soon, perhaps should change the order of the text?

I think more info here is better than "go search yourself", because this project could be used by very beginner users, and it has said the gpg stuff is PITA even for novice users.

Ocramius replied:
Would be best to not recommend a value nor provide any guidance for the key length at all: this target changes, and the first time a security researcher with a bad mood shows up we have a problem :-)

heiglandreas replied:
For the same reasons I would not recommend any cryptosystem here. Whether people use RSA or DSA or any ElipticCurve or whatnot  is their preference and dependend on their toolset.

Ocramius replied:
Aye: the only requirement is that the key must be usable for signing :+1:

glensc replied:
Ok. let's remove any recommendation then, just "chose signing"

glensc committed:
Remove any type and bits recommendation

Signed-off-by: Elan Ruusamäe <glen@pld-linux.org>

glensc commented:
@Ocramius I have nothing more to add here at this point.

glensc commented:
Another ping!

Ocramius reviewed (approved):
@glensc thanks for improving the docs here - sorry for the massive delay!

glensc commented:
@Ocramius sure, we have our own distraction demons. I planned to check out the "new key" section too but never got there.
